dbms/src/Interpreters/Join.cpp

```diff
@@ -1529,6 +1532,18 @@ Block doNASemiJoinOrSemiJoin(
         if (is_cancelled())
             return {};
         helper->doJoin();
+        // because pipeline model is enabled by default, current_task_timer should be not null in most senarios
+        if likely (current_task_timer != nullptr)
+        {
+            auto elapsed = current_task_timer->updateExecutingTime();
+            auto time_exceed = elapsed >= 6 * YIELD_MAX_TIME_SPENT_NS;
+            fiu_do_on(FailPoints::force_semi_join_time_exceed, { time_exceed = true; });
+            if unlikely (time_exceed)
+            {
+                // task execution time exceeds the maximum time, yield the thread by returning an empty result block
+                return empty_result_block;
+            }
+        }
     }
 
     if (is_cancelled())
```

Comment:
Move the time_exceed check after is_cancelled() check?

Reply:
`time_exceed` check need to be done inside the `while` loop because all the complex calculation is in the `while` loop